File: trees/js/BinaryNode.js
/* This source file contains binary nodes, intended to be used by binary search
 * trees. They contain one value of any type and slots to link to left or right
 * nodes. A possible extension includes adding a link to a parent node. */

module.exports = function BinaryNode(value, left, right) {
  this.val = value;
  this.left = left;
  this.right = right;
}

File: trees/java/BinaryNode.java
package io.github.afresnedo.trees;

// A generic binary node, without a link to its parent, that contains
// comparable data

class BinaryNode<T extends Comparable<? super T>> {

    T val;
    BinaryNode<T> left;
    BinaryNode<T> right;

    public BinaryNode(T value) {
        this(value, null, null);
    }

    public BinaryNode(T value, BinaryNode<T> leftChild,
            BinaryNode<T> rightChild) {
        this.val = value;
        this.left = leftChild;
        this.right = rightChild;
    }
}
